File: typescript/stack/stack.ts
interface Stack<T> {
    push(item: T): void;
    pop(): T | undefined;
    peek(): T | undefined;
    isEmpty(): boolean;
    isFull(): boolean;
    printStack(): void;
}

class SimpleStack<T> implements Stack<T> {
    private readonly capacity: number;
    private items: T[] = [];

    constructor(capacity: number = 5) {
        this.capacity = capacity;
    }

    push(item: T): void {
        if (this.isFull()) {
            console.log("Stack Overflow");
            return;
        }
        this.items.push(item);
    }

    pop(): T | undefined {
        if (this.isEmpty()) {
            console.log("Stack Underflow");
            return undefined;
        }
        return this.items.pop();
    }

    peek(): T | undefined {
        if (this.isEmpty()) {
            return undefined;
        }
        return this.items[this.items.length - 1];
    }

    isEmpty(): boolean {
        return this.items.length === 0;
    }

    isFull(): boolean {
        return this.items.length === this.capacity;
    }

    printStack(): void {
        console.log(this.items);
    }
}

export default SimpleStack;

File: zig/stack/src/stack.zig
const std = @import("std");

pub fn Stack(comptime T: type, comptime capacity: usize) type {
    return struct {
        items: [capacity]T,
        top: usize,

        const Self = @This();

        pub fn init() Self {
            return Self{
                .items = undefined,
                .top = 0,
            };
        }

        pub fn push(self: *Self, value: T) !void {
            if (self.top == capacity) {
                return error.StackOverflow;
            }
            self.items[self.top] = value;
            self.top += 1;
        }

        pub fn pop(self: *Self) !T {
            if (self.top == 0) {
                return error.StackUnderflow;
            }
            self.top -= 1;
            return self.items[self.top];
        }

        pub fn peek(self: *const Self) ?T {
            if (self.top == 0) {
                return null;
            }
            return self.items[self.top - 1];
        }

        pub fn isEmpty(self: *const Self) bool {
            return self.top == 0;
        }

        pub fn isFull(self: *const Self) bool {
            return self.top == capacity;
        }

        pub fn printStack(self: *const Self) void {
            std.debug.print("Stack: [", .{});
            var i: usize = 0;
            while (i < self.top) : (i += 1) {
                std.debug.print("{}, ", .{self.items[i]});
            }
            std.debug.print("]\n", .{});
        }

        pub fn deinit(self: *Self) void {
            self.top = 0;
            for (self.items[0..capacity]) |*item| {
                item.* = undefined;
            }
        }
    };
}

pub const StackError = error{
    StackOverflow,
    StackUnderflow,
};
